### PR TITLE
feat(swap): integrate ellipsis finance swap

### LINKED ISSUE #5792: [Bounty] Integrate Ellipsis, A Secure Low-slippage Stable Swapping on BSC into Mask Plugin
## Catalogue

- Platform
  - [x] Gitcoin
  - [ ] Hackerlink
  - [ ] Mask
- Experience
  - [ ] Beginner
  - [x] Intermediate
  - [ ] Advanced
- Category
  - [ ] Mask Core Feature
  - [x] Mask Plugin
  - [ ] Security/Cryptography Feature
  - [ ] Documentation
  - [ ] Design
  - [ ] Other
    - [ ] /_List it here_/
- Time Commitment
  - [x] 2 weeks 
- Bounty Size
  - [x] 1000 USDC or $Mask equivalent 

## Part 1: Problem Description

**Background**  Ellipsis.finance is an exchange for stable coins on the Binance Smart Chain (BSC). It allows for stable coins swap with very low slippage and minimal fees. Ellipsis is an authorised fork of Curve Finance. As a result of this agreement, veCRV holders will receive 25% of the supply over the next 12 months.

Curve Finance will provide support to the Ellipsis team. We have committed to Curve core values of which are :
a. no deposit or withdrawal fees,
b. no lock ups on liquidity, and
c. extremely efficient stable coin swaps.

**Utility**
The Ellipsis token (EPS) is a revenue earning token. EPS stakers can earn fees from the Ellipsis protocol.

**Trading Fees**
Fees will be equally split between EPS stakers and liquidity providers (50%/50%).

**Solution** 
Create a plugin that integrates Ellipsis into Mask. The UI should be identical to other swaps

image

image

## Part 2: Resources

- [Bounty Guide](https://github.com/DimensionDev/Maskbook/tree/master/docs/bounty-guide.md)
- https://ellipsis.finance/
- https://docs.ellipsis.finance/
- https://github.com/ellipsis-finance

## Part 3: Contact and Support
@yisiliu @tzhan28 @LindaMask for ideas
@Jack-Works for implementations
Mask Network discord server: https://discord.gg/4Xccj7a75Q,
Ellipsis Telegram: https://t.me/ellipsisfinance

Please speak up in the Mask discord's bounty section or just discord dm @yisiliu or @taylor zhang or @lin_dao #5792 so that we can find you on Discord. For faster communication reach out to lin_DAO on telegram and she will create a dedicated group with the Mask team to start the bounty process

Submission Requirements
Please include demo to show the integration is working.

Judging Criteria
Have a working demo.
Clean code & tech design
Clear UI, easy to use and generally good user experience

### PR DESCRIPTION
**Pull request by brymut**
## Description

Integrate Ellipsis Finance for Swapping in Mask Plugin on BSC.

Closes #5792 

## Type of change

- [x] New feature (non-breaking change which adds functionality)

## Checklist

- [x] My code follows the style guidelines of this project.
- [x] I have performed a self-review of my own code.
  - [x] I have removed all in development `console.log`s
  - [x] I have removed all commented code.
- [x] I have commented on my code, particularly in hard-to-understand areas.
- [x] I have read [Internationalization Guide](https://github.com/DimensionDev/Maskbook/blob/develop/docs/i18n-guide.md) and moved text fields to the i18n JSON file.

### CODE REVIEW

File: packages/web3-constants/evm/trader.json [hunk 1]
```diff
@@ -1018,5 +1018,85 @@
...
+        "Arbitrum_Rinkeby": "",
+        "xDai": "",
+        "Avalanche": "",
+        "Avalanche_Fuji": "",
+        "Celo": "",
+        "Fantom": "",
+        "Aurora": "",
+        "Aurora_Testnet": ""
+    },
+    "ELLIPSIS_INIT_CODE_HASH": {
```
Comment (guanbinrui) [inline packages/web3-constants/evm/trader.json:1151]: Plz complete this field.

Comment (brymut) [reply packages/web3-constants/evm/trader.json:1151]: apologies, still looking for that information, not available on ellipsis docs

Comment (guanbinrui) [reply packages/web3-constants/evm/trader.json:1151]: You can find it from the source code.

Comment (brymut) [reply packages/web3-constants/evm/trader.json:1151]: Their code is not opensource, I've been trying to be in contact with their devs from their telegram but still waiting for responses, might you be able to point me in the right direction of where I could look?

**Comment by UncleBill** [comment]
image

@brymut , You need to make an option of Ellipsis add to the trader list here. The best practice is add correct contract addresses of the dex in https://github.com/DimensionDev/Maskbook/blob/793a1b355cfb5584dc35932d3ef706e6f0e9609a/packages/web3-constants/evm/trader.json

**Comment by brymut** [comment]
> 
> @brymut , You need to make an option of Ellipsis add to the trader list here. The best practice is add correct contract addresses of the dex in https://github.com/DimensionDev/Maskbook/blob/793a1b355cfb5584dc35932d3ef706e6f0e9609a/packages/web3-constants/evm/trader.json

Yes, its just that I'm having trouble finding that data, I'm still waiting for the ellipsis team to get back to me since their codebase is not opensource

**Comment by UncleBill** [comment]
> Yes, its just that I'm having trouble finding that data, I'm still waiting for the ellipsis team to get back to me since their codebase is not opensource

You can try tracing a TX or from their frontend.

**Comment by brymut** [comment]
I have also been working on https://github.com/DimensionDev/Maskbook/pull/5863 which I've had better progress on if you can have a look, almost done but needs some more changes.

**Comment by guanbinrui** [comment]
Hi @brymut , I just found that the ellipsis finance swap is not a fork of Uniswap. But a fork from Curve.fi.

+ [curve contract](https://github.com/curvefi/curve-contract/tree/master/contracts)
+ [ellipsis-finanace contracts](https://github.com/ellipsis-finance/ellipsis/tree/master/contracts)

It means you cannot implement it by extending from the Uniswap protocol. So, Plz, read the [doc](https://curve.readthedocs.io/exchange-pools.html) to understand curve.fi swap protocol.

**Comment by guanbinrui** [comment]
@brymut Plz resolve those conflicts.

**Comment by Jack-Works** [comment]
Please fix merge conflicts and type errors.

**Pull request closed without merging**